### PR TITLE
Feat/taro-update-backend

### PR DESCRIPTION
**Pull request by amovfx**
Create a draft of this for you to inspect workflow. I'm still working on tests and coverage. Also, compatibility filtering isn't working. I will figure that out.

### Description

Modal for changeing the backend of taro nodes.

### Steps to Test

1. Create 2 LND nodes and 1 Taro Node
2. Click the link between the taro node and lnd node
3. Click 'Change Backend' on info panel.
4. Change the lnd node selection.
5. Click 'Change Backend' button.

### CODE REVIEW

File: src/components/designer/taro/actions/ChangeTaroBackendModal.tsx [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+  useEffect(() => {
+    const { lightning, taro } = network.nodes;
+    const taroNode = taro.find(n => n.name === selectedTaro);
+    const lndBackend = lightning.find(n => n.name === selectedLNDBackend);
+    if (taroNode && lndBackend) {
+      const { compatibility } = dockerRepoState.images[taroNode.implementation];
+      if (compatibility) {
+        const requiredVersion = compatibility[lndBackend.version];
+        //version compatibility function breaks down
+        if (!requiredVersion) {
```
Comment (amovfx) [inline src/components/designer/taro/actions/ChangeTaroBackendModal.tsx:73]: I had a lot of trouble hitting this if statement, I couldn't figure out how what test you were running where your coverage would clear.

**Review by amovfx — COMMENTED** [review]
This has been rebased to the latest add-taro. 

100% Coverage and linting checked.

**Comment by jamaljsr** [comment]
This needs a rebase on master since #656 and #667 were merged. One tip to make rebases a little easier is to do an interactive rebase then drop all of the commits that are already in the base branch.

```sh
git rebase -i feat/add-taro
```

I did a pass on this to test the UX. I noticed some things to improve:
1. You should be able to drag from the bottom port of a taro node to the top port of an LND node to open the change backend modal. 
2. You shouldn't be able to change the backend if the taro node has assets. Those assets are embedded in UTXOs in the wallet of the LND node it is currently connected to. If you change the backend, then the taro node will no longer have access to the UTXOs in the LND wallet so those assets are essentially burned. The challenge is going to be how to tell if the taro node has assets if changing the backend when the network is stopped. There may be some data on disk that you can check. I'm not sure. This would require a bunch of testing scenarios and/or asking questions in the LND slack #taro channel about what users should & shouldn't do when changing nodes.

**Review by amovfx — COMMENTED** [review]
I think this might be the last feature on the addTaro check list.

Coverage, testing and linting 100%.

File: src/components/designer/NetworkDesigner.spec.tsx [hunk 2]
```diff
@@ -147,6 +147,15 @@ describe('NetworkDesigner Component', () => {
...
   });
+  it('should display the ChangeTaroBackend modal', async () => {
+    const { findAllByText, findByText, getAllByText, store } = renderComponent();
+    expect(await findByText('backend1')).toBeInTheDocument();
+    act(() => {
+      store.getActions().modals.showChangeTaroBackend({});
+    });
+    expect(await findAllByText('Change Taro Node Backend')).toHaveLength(2);
+    fireEvent.click(getAllByText('Cancel')[0]);
+  });
```
[the remark this answers is not in the document]
Comment (jamaljsr) [reply src/components/designer/NetworkDesigner.spec.tsx:158]: This test is passing in the commits currently on Github. Did you forget to force-push your rebase?

[the remark this answers is not in the document]
Comment (amovfx) [reply src/components/designer/NetworkDesigner.spec.tsx:158]: I think that is the mistake I am making.

**Comment by amovfx** [comment]
I added the drag link feature. 

`You shouldn't be able to change the backend if the taro node has assets. Those assets are embedded in UTXOs in the wallet of the LND node it is currently connected to. If you change the backend, then the taro node will no longer have access to the UTXOs in the LND wallet so those assets are essentially burned. The challenge is going to be how to tell if the taro node has assets if changing the backend when the network is stopped. There may be some data on disk that you can check. I'm not sure. This would require a bunch of testing scenarios and/or asking questions in the LND slack #taro channel about what users should & shouldn't do when changing nodes.`

Thanks for clarifying this behavior. I'll look into a solution. Sounds like a good challenge.

**Comment by amovfx** [comment]
K this is ready for review.

**Review by jamaljsr — COMMENTED** [review]
Did another pass of the UX and found some issues. Can you resolve these before I dive into code review?

- With `carol-taro` connected to `carol`, drag a new link between the two and click Change Backend. It will display an already connected error, which is correct. But now if you click on the link, the sidebar says invalid selection. You will also see a link remains if you drag from `alice-taro` to `bob` then hit Cancel in the modal. You have to remove the temporary link that gets added when dragging.
- I am allowed to change the backend of a started network by dragging a link between the nodes. It should display an error
- With the network started, the Change Taro backend modal displays a message stating the node will be restarted automatically. This msg should be removed since we won't allow changing after initial startup. Also remove the code that restarts the node
- I'm getting failing tests in `LinkDetails.spec.tsx` due to this error
  ```
  /Users/jamal/dev/polar/src/utils/config.ts:208
    join)(network.path, 'volumes',
                  ^

  TypeError: Cannot read properties of undefined (reading 'path')
  ```
  Note: to make your life easier, it would be fine to ignore the unit tests for now. When the UX/code are approved and final, then you can write all of the test. I usually implement the full feature first, then write tests at the end, so I don't have to repeated rewrite the tests over and over again while iterating on the code.

**Comment by amovfx** [comment]
Thanks for the feedback and guidance, I'll get on this.

**Comment by amovfx** [comment]
I hit these notes, thanks for running through this, this has been a challenging task.

File: src/i18n/locales/en-US.json [hunk 3]
```diff
@@ -118,6 +118,9 @@
   "cmps.designer.link.Backend.version": "Version",
   "cmps.designer.link.Backend.status": "Status",
   "cmps.designer.link.ChangeBackendButton.btnText": "Change Backend",
+  "cmps.designer.link.ChangeTaroBackendButton.btnText": "Change Taro Backend",
+  "cmps.designer.link.ChangeTaroBackendButton.error": "Error",
```
Comment (jamaljsr) [inline src/i18n/locales/en-US.json:122]: Remove because this isn't used by the ChangeTaroBackendButton component

File: src/i18n/locales/en-US.json [hunk 4]
```diff
@@ -147,6 +150,14 @@
...
   "cmps.designer.link.Peer.status": "Status",
+  "cmps.designer.link.lndbackend.title": "Taro Lnd Connection",
+  "cmps.designer.link.lndbackend.desc": "Taro nodes communicate over lnd to mint and transfer assets.",
+  "cmps.designer.link.lndbackend.taroTitle": "Taro Node",
+  "cmps.designer.link.lndbackend.lndTitle": "Lnd Node",
+  "cmps.designer.link.lndbackend.name": "Name",
+  "cmps.designer.link.lndbackend.implementation": "Implementation",
+  "cmps.designer.link.lndbackend.version": "Version",
+  "cmps.designer.link.lndbackend.status": "Status",
   "cmps.designer.link.TaroBackend.title": "Taro Backend Connection",
```
Comment (jamaljsr) [inline src/i18n/locales/en-US.json:161]: These are not used by any component.

File: src/i18n/locales/en-US.json [hunk 5]
```diff
@@ -176,6 +187,17 @@
   "cmps.designer.lightning.actions.ChangeBackendModal.compatWarning": "{{ln.name}} is running {{ln.implementation}} v{{ln.version}} which is compatible with Bitcoin Core v{{requiredVersion}} and older. {{backend.name}} is running v{{backend.version}} so it cannot be used.",
   "cmps.designer.lightning.actions.ChangeBackendModal.successTitle": "Successfully updated the backend node",
   "cmps.designer.lightning.actions.ChangeBackendModal.successDesc": "The {{ln}} node will pull chain data from {{backend}}",
+  "cmps.designer.lightning.actions.ChangeTaroBackendModal.title": "Change Taro Node Backend",
+  "cmps.designer.lightning.actions.ChangeTaroBackendModal.description": "Changing the Taro Node's backend will instruct it to use a different LND node to send/mint assets.",
+  "cmps.designer.lightning.actions.ChangeTaroBackendModal.taroNodeLabel": "Taro Node",
+  "cmps.designer.lightning.actions.ChangeTaroBackendModal.LNDNodeLabel": "LND Node",
+  "cmps.designer.lightning.actions.ChangeTaroBackendModal.restartNotice": "The {{name}} node will be restarted automatically to apply the change.",
```
Comment (jamaljsr) [inline src/i18n/locales/en-US.json:194]: This isn't used by the ChangeTaroBackendModal component.

File: src/components/designer/taro/actions/ChangeTaroBackendModal.tsx [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+          setCompatWarning(
+            l('compatWarning', {
+              taroNode,
+              lndBackend,
+              requiredVersion: '2022.12.28-master',
```
Comment (jamaljsr) [inline src/components/designer/taro/actions/ChangeTaroBackendModal.tsx:73]: This version shouldn't be hard-coded. Did you mean to use `requiredVersion`?

File: src/i18n/locales/en-US.json [hunk 7]
```diff
@@ -478,6 +500,9 @@
   "store.models.designer.linkErrPorts": "Use the top & bottom ports to connect between bitcoin and lightning nodes",
   "store.models.designer.dropErrTitle": "Failed to add node",
   "store.models.designer.dropErrMsg": "Cannot add a new node when the network is transitioning",
+  "store.models.designer.linkErrTarod": "Taord nodes do not connect",
```
Comment (jamaljsr) [inline src/i18n/locales/en-US.json:503]: Change to "Taro nodes do not connect to eachother"

File: src/store/models/designer.ts [hunk 8]
```diff
@@ -181,6 +184,24 @@ const designerModel: DesignerModel = {
       },
     };
   }),
+  updateTaroBackendLink: action((state, { taroName, lndName }) => {
+    const chart = state.allCharts[state.activeId];
+    // remove the old ln -> backend link
```
Comment (jamaljsr) [inline src/store/models/designer.ts:189]: This should be "old taro -> LN link"

File: src/store/models/designer.ts [hunk 9]
```diff
@@ -245,6 +266,31 @@ const designerModel: DesignerModel = {
       } else if (fromNode.type === 'bitcoin' && toNode.type === 'bitcoin') {
         // connecting bitcoin to bitcoin isn't supported
         return showError(l('linkErrBitcoin'));
+      } else if (fromNode.type === 'taro' || toNode.type === 'taro') {
+        if (fromNode.type === 'taro' && toNode.type === 'taro') {
+          //connecting taro to taro isn't supported
+          return showError(l('linkErrTarod'));
+        }
+        const taroName = fromNode.type === 'taro' ? fromNodeId : toNodeId;
+        const lndeName = fromNode.type === 'taro' ? toNodeId : fromNodeId;
```
Comment (jamaljsr) [inline src/store/models/designer.ts:275]: typo: should be `lndName`

File: src/store/models/designer.ts [hunk 10]
```diff
@@ -245,6 +266,31 @@ const designerModel: DesignerModel = {
...
+        const lnNetworkNode = network.nodes.lightning.find(
+          n => n.name === lndeName && n.implementation === 'LND',
+        ) as LndNode;
+        if (!lnNetworkNode) {
+          return showError(l('linkErrLNDImplementation', { nodeName: lndeName }));
+        }
+        const taroData = getTarodFilePaths(taroName, network);
```
Comment (jamaljsr) [inline src/store/models/designer.ts:282]: You don't need to call `getTarodFilePaths` because the path is already defined on the node in `network.nodes.taro[]`. The `node.path.adminMacaroon` field.

File: src/store/models/designer.ts [hunk 11]
```diff
@@ -245,6 +266,31 @@ const designerModel: DesignerModel = {
...
+        exists(taroData.adminMacaroon).then(result => {
```
Comment (jamaljsr) [inline src/store/models/designer.ts:283]: Try to stay away from `.then()`. Use `await` instead when possible.

File: src/store/models/modals.ts [hunk 12]
```diff
@@ -81,6 +88,7 @@ export interface ModalsModel {
   assetInfo: AssetInfoModel;
   mintAsset: MintAssetModel;
   newAddress: NewAddressModel;
+  changeTaroBackend: ChangeTaroBackendModel;
```
Comment (jamaljsr) [inline src/store/models/modals.ts:91]: nit: to be more organized in this file, keep the order of the types/fields/actions consistent. Place all the new stuff at the botton of their groups.

File: src/store/models/network.spec.ts [hunk 13]
```diff
@@ -495,6 +504,53 @@ describe('Network model', () => {
     });
   });
 
+  describe('Updating Taro Backend', () => {
+    beforeEach(async () => {
+      createTaroNetwork();
+    });
```
Comment (jamaljsr) [inline src/store/models/network.spec.ts:510]: Place these tests in the `describe("Taro network")` group to eliminate the need for the `createTaroNetwork()` func.

File: src/store/models/network.ts [hunk 14]
```diff
@@ -107,6 +108,12 @@ export interface NetworkModel {
     StoreInjections,
     RootModel
   >;
+  updateTaroBackendNode: Thunk<
+    NetworkModel,
+    { id: number; taroName: string; LNDName: string },
```
Comment (jamaljsr) [inline src/store/models/network.ts:113]: `LNDName` should be camel case `lndName`. `LND` should only be all caps when displaying the product name to users. All caps should only be used in code when defining a type that should start with a capital letter (ex: `class LNDService`). Variable names should always start with lower-case, so `lnd` is correct.

File: src/store/models/network.ts [hunk 15]
```diff
@@ -508,6 +515,36 @@ const networkModel: NetworkModel = {
...
+        throw new Error(l('connectedErr', { lnName: taroName, backendName: LNDName }));
+
+      taroNode.lndName = lndNode.name;
+      // update the network in the redux state and save to disk
+      actions.setNetworks([...networks]);
+      await actions.save();
+      // save the updated compose file
+      await injections.dockerService.saveComposeFile(network);
+
+      if (network.status === Status.Stopped) {
```
Comment (jamaljsr) [inline src/store/models/network.ts:541]: It doesn't look like you need this check here since this action is never called when the network isn't stopped.

File: src/utils/tests/helpers.ts [hunk 16]
```diff
@@ -148,7 +151,7 @@ export const getNetwork = (
     lndNodes: 2,
     clightningNodes: 1,
     eclairNodes: 1,
-    bitcoindNodes: 1,
+    bitcoindNodes: 2,
```
Comment (jamaljsr) [inline src/utils/tests/helpers.ts:154]: It's not clear to me why you needed to increase the number of bitcoin nodes for the tests. Why was this necessary?

**Review by jamaljsr — COMMENTED** [review]
Functionality looks good now. 

Did a first pass of code review.

**Comment by amovfx** [comment]
This should be good for review.

File: src/components/designer/link/ChangeTaroBackendButton.tsx [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+    } else {
+      showChangeTaroBackend({ taroName, lndName });
+    }
+  };
+
+  useEffect(() => {
+    (async () => {
+      const taroNode = network.nodes.taro.find(n => n.name === taroName) as TarodNode;
+      const result = await exists(taroNode.paths.adminMacaroon);
+      setHasMacaroon(result);
```
Comment (jamaljsr) [inline src/components/designer/link/ChangeTaroBackendButton.tsx:45]: You can run this code in the `handleChangeClick` callback instead of adding a state var.

File: src/i18n/locales/en-US.json (hunk 3 above)
Comment (jamaljsr) [reply src/i18n/locales/en-US.json:122]: Please see my prior comments about changes to this file. They haven't been updated in your latest commits.

File: src/store/models/modals.ts [hunk 18]
```diff
@@ -16,6 +16,13 @@ interface ChangeBackendModel {
   linkId?: string;
 }
 
+interface ChangeTaroBackendModel {
```
Comment (jamaljsr) [inline src/store/models/modals.ts:19]: nit: this declaration should be below `SendAssetModal`.

File: src/store/models/modals.ts [hunk 19]
```diff
@@ -292,6 +308,23 @@ const modalsModel: ModalsModel = {
       ...payload,
     };
   }),
+  showChangeTaroBackend: thunk((actions, { taroName, lndName, linkId }) => {
```
Comment (jamaljsr) [inline src/store/models/modals.ts:311]: nit: These 3 new functions should be placed below the `*SendAsset` functions.

File: src/store/models/network.spec.ts [hunk 20]
```diff
@@ -48,6 +48,15 @@ describe('Network model', () => {
   // helper to get the first network in the store
   const firstNetwork = () => store.getState().network.networks[0];
 
+  const createTaroNetwork = () => {
+    const network = getNetwork(1, 'test network', Status.Stopped, 2);
```
Comment (jamaljsr) [inline src/store/models/network.spec.ts:52]: I see you moved the test cases into the `Taro network` group. No you no longer need the `createTaroNetwork` function declared here. You can just move this code back into the `beforeEach`.

File: src/utils/tests/helpers.ts (hunk 16 above)
Comment (jamaljsr) [reply src/utils/tests/helpers.ts:154]: Do we need this change?

File: src/i18n/locales/en-US.json [hunk 21]
```diff
@@ -147,6 +151,14 @@
   "cmps.designer.link.Peer.implementation": "Implementation",
   "cmps.designer.link.Peer.version": "Version",
   "cmps.designer.link.Peer.status": "Status",
+  "cmps.designer.link.lndbackend.title": "Taro Lnd Connection",
```
Comment (jamaljsr) [inline src/i18n/locales/en-US.json:154]: These `lndbackend` keys aren't used by any component and can be removed.

File: src/i18n/locales/en-US.json [hunk 22]
```diff
@@ -176,6 +188,17 @@
   "cmps.designer.lightning.actions.ChangeBackendModal.compatWarning": "{{ln.name}} is running {{ln.implementation}} v{{ln.version}} which is compatible with Bitcoin Core v{{requiredVersion}} and older. {{backend.name}} is running v{{backend.version}} so it cannot be used.",
   "cmps.designer.lightning.actions.ChangeBackendModal.successTitle": "Successfully updated the backend node",
   "cmps.designer.lightning.actions.ChangeBackendModal.successDesc": "The {{ln}} node will pull chain data from {{backend}}",
+  "cmps.designer.lightning.actions.ChangeTaroBackendModal.title": "Change Taro Node Backend",
+  "cmps.designer.lightning.actions.ChangeTaroBackendModal.description": "Changing the Taro Node's backend will instruct it to use a different LND node to send/mint assets.",
+  "cmps.designer.lightning.actions.ChangeTaroBackendModal.taroNodeLabel": "Taro Node",
+  "cmps.designer.lightning.actions.ChangeTaroBackendModal.LNDNodeLabel": "LND Node",
+  "cmps.designer.lightning.actions.ChangeTaroBackendModal.restartNotice": "The {{name}} node will be restarted automatically to apply the change.",
```
Comment (jamaljsr) [inline src/i18n/locales/en-US.json:195]: `restartNotice` isn't used by the `ChangeTaroBackendModal` component.

**Review by jamaljsr — COMMENTED** [review]
Thanks for the updates. I just reiterated a couple comments on  the `en-US.json` file. Then this should be good to go.

**Comment by amovfx** [comment]
Good to go.

**Review by jamaljsr — APPROVED** [review]
Perfect 👌

Thanks for iterating on this. Going to squash merge this because there are a lot of commits and the diff is not that big.
